app/pages/learn/01_tutorial/01_your-first-java-app/02_building-with-intellij-idea.md

```diff
@@ -0,0 +1,262 @@
+        
+        double actualAverage = Average.calculateAverage(numbers);
+        
+        double expectedAverage = 3.0;
+        assertEquals(expectedAverage, actualAverage);
+    }
+
+    @Test
+    void shouldReturnZeroAverageForEmptyArray() {
+        int[] numbers = {};
+        
+        double actualAverage = Average.calculateAverage(numbers);
+        
+        double expectedAverage = 0;
+        assertEquals(expectedAverage, actualAverage);
+    }
+}
+```
+In our test class, we can select **Run All Tests** (**⌃⇧R** on macOS or **Ctrl+Shift+F10** on Windows/Linux).
+
+In our example, we see that the second tests fails. We expected to get the value `0` as the average of an empty array, but got `NaN` (not a number) instead. Let's find out why, using the debugger.
+
+## Debugging
+We might want to see how our code runs, either to help us understand how it works and/or when we need to fix a bug or failing test, like the one above. We can run our code through the [debugger](https://www.jetbrains.com/help/idea/debugging-code.html) to see the state of our variables at different times, and the call stack - the order in which methods are called when the program executes. To do so, we must first add a [breakpoint](https://www.jetbrains.com/help/idea/using-breakpoints.html) to the code.
+
+To add a breakpoint, click the gutter at the line of code where you want execution to stop. Alternatively, place the caret at the line and press **⌃F8** (on macOS) or **Ctrl+F8** (on Windows/Linux). We can run our test or application using the **Debug** option.
+
+Execution will stop at the breakpoint, so we can investigate the state of our application. We can see current values of variables and objects. We can evaluate an expression, to see its current value and look at more details. We can even change the expressions to evaluate different results. We can continue execution by either stepping into a method to see what happens inside a called method (using the shortcut **F7**, or the corresponding button in the *Debug* tool window) or stepping over a line to go to the next line even if a method is called (using the shortcut **F8**, or the corresponding button in the *Debug* tool window), depending on what we’re interested in. Finally, we can resume the program to finish the execution of the test.
+
+Let's debug the failing test from the previous section. In the code, place a breakpoint on line 4. Run the failing test through the debugger. Step over the code until you get to line 8, and observe the values of the variables. When we get to line 8, select `sum / numbers.length`, right-click to open the context menu and select **Evaluate Expression**. Press **Enter** to evaluate the selected expression. We see that `sum / numbers.length` results in a `java.lang.ArithmeticException: / by zero`. The empty array has a length of `0` and Java does not allow dividing by zero. When we try to cast this Exception to a `(double)` we get `NaN`; an exception is not a number.
```

Comment:
Nothing is _casting an exception_ here.
> When we try to cast this Exception to a `(double)` we get `NaN`; an exception is not a number.

makes no sense.

Reply:
Why do we get a NaN here?

Reply:
What do you mean with this question? I took that quote from your article. But `NaN` would the result of a division of zero by zero if one of the _arguments_ was a IEEE754 floating point number so I guess it's supposed to mean that one of the arguments would need to be a `double` in order to not get an exception.
However, I also think that `NaN` is not really a good value to deal with.

Reply:
I didn't get your first comment either.

Reply:
The article contains the sentence
> When we try to cast this Exception to a `(double)` we get `NaN`; an exception is not a number.

However, it is not the `Exception` being cast to `double` but one of the operands that would need to be cast to a `double` in order to avoid an exception.
That being said, I don't think returning `NaN` in a method like that would be a good idea anyways.

Reply:
To solve this conversation, @mlvandijk , how about changing the last two phrases to 
_The empty array has a length of 0 and Java does not allow dividing by zero. hence it returns the dataype `NaN`(NotaNumber)._